src/MagentoHackathon/Composer/Magento/Patcher/Bootstrap.php

```diff
@@ -24,86 +55,158 @@ private function getConfig()
     }
 
+    /**
+     * @return string
+     */
     protected function getAppPath()
     {
         return $this->getConfig()->getMagentoRootDir() . '/app';
     }
 
-    protected function splitOriginalMage()
+    /**
+     * @return bool
+     */
+    protected function writeComposerAutoloaderPatch()
     {
-        $appPath = $this->getAppPath();
-        if (file_exists($appPath . '/Mage.class.php')) {
-            return;
-        }
+        $mageFileContent = file($this->getMageClassFilePath());
 
-        $mageFileContent = file($appPath . '/Mage.php');
+        $mageFileBootstrapPart = '';
+        $mageFileClassDeclarationPart = '';
+        $isBootstrapPart = true;
 
-        $mageClassFile = '';
-        $mageBootstrapFile = '<?php' . PHP_EOL;
-        $isBootstrapPart = false;
         foreach ($mageFileContent as $row) {
-            if (strpos($row, 'define') === 0) {
-                $isBootstrapPart = true;
-            }
             if ($isBootstrapPart) {
-                $mageBootstrapFile .= $row;
+                $mageFileBootstrapPart .= $row;
             } else {
-                $mageClassFile .= $row;
+                $mageFileClassDeclarationPart .= $row;
             }
             if (strpos($row, 'Varien_Autoload') === 0) {
                 $isBootstrapPart = false;
             }
         }
-        $mageClassFile .= PHP_EOL;
-        $mageBootstrapFile .= PHP_EOL;
-        file_put_contents($appPath . '/Mage.class.php', $mageClassFile);
-        file_put_contents($appPath . '/Mage.bootstrap.php', $mageBootstrapFile);
 
-        $mageFileReplacement
-            = <<<php
-<?php
-require __DIR__ . '/bootstrap.php';
+        $mageFileReplacement = $mageFileBootstrapPart . PHP_EOL
+                             . $this->getAutoloaderPatchString() . PHP_EOL
+                             . $mageFileClassDeclarationPart;
 
-php;
-        file_put_contents($appPath . '/Mage.php', $mageFileReplacement);
+        return file_put_contents($this->getMageClassFilePath(), $mageFileReplacement) !== false;
+    }
 
+    /**
+     * @param IOInterface $io
+     */
+    public function setIo(IOInterface $io)
+    {
+        $this->io = $io;
+    }
 
+    /**
+     * @return IOInterface
+     */
+    public function getIo()
+    {
+        if (!$this->io) {
+            $this->io = new NullIO();
+        }
+        return $this->io;
     }
 
-    protected function generateBootstrapFile()
+    /**
+     * @return string
+     */
+    private function getAutoloaderPatchString()
     {
-        $appPath = $this->getAppPath();
-        $bootstrapFile
-            = <<<php
-<?php
-require __DIR__ . '/Mage.class.php';
-require __DIR__ . '/Mage.bootstrap.php';
+        $patchMark = self::PATCH_MARK;
+        return <<<PATCH
+/** $patchMark **/
+\$autoloaderPath = '{$this->getVendorAutoloaderPath()}';
+if (file_exists(\$autoloaderPath)) {
+    require_once \$autoloaderPath;
+}
```

Comment:
The autoloader path if pretty much guaranteed (as much as you can with composer), because that path comes from here https://github.com/inviqa/magento-composer-installer/pull/3/files#diff-696ef84c63e9db98ae0c3ee424cfceb5R210 - aka directly from Composer's innards about the vendor-dir path.